File: leetcode/tree/230. Kth Smallest Element in a BST/js/main.js
// DFS
// Time O(N)
// Space O(N)
const kthSmallest = (root, k) => {
  let stack = [];

  while (stack.length || root != null) {
    while (root != null) {
      stack.push(root);
      root = root.left;
    }

    root = stack.pop();
    k--;

    if (k == 0) {
      return root.val;
    }

    root = root.right;
  }

  return null;
};

File: leetcode/tree/230. Kth Smallest Element in a BST/py/main.py
# Time O(N)
# Space O(N)
class Solution:
  def kthSmallest(self, root, k):
    st = [root]
    
    while True:
      while root:
        st.append(root)
        root = root.left
      
      root = st.pop()
      k -= 1
      
      if not k:
        return root.val
      
      root = root.right